### PR TITLE
Bug 1381869 - Use separate elasticsearch index for Bugzilla::User

### PR DESCRIPTION
**Pull request by dylanwh**
This patch removes the concept of a single, bugzilla-wide index in favor of a per-class index.
bugs and comments continue to use Bugzilla->params->{elasticsearch_index} but users
use Bugzilla->params->{elasticsearch_index} . "_users".

It is assured via the ChildObject trait (role) that comments will share the index with bugs, and we have kept the index for bugs/comments the same to avoid the multi-hour reindexing of production. Re-indexing users takes only five minutes.

Subsequent work on this will allow use to version the index names and use aliases, but I wanted to keep this patch small.

This patch also corrects a mistake: $indexer->put_mapping() should not have been a public method.

### CODE REVIEW

File: Bugzilla/Bug.pm [hunk 1]
```diff
@@ -302,6 +302,73 @@ with 'Bugzilla::Elastic::Role::Object';
 
 sub ES_TYPE {'bug'}
 
+sub ES_INDEX { Bugzilla->params->{elasticsearch_index} }
+
+sub ES_SETTINGS {
```
Comment (floatingatoll) [inline Bugzilla/Bug.pm]: Were these patterns copied as-is from their prior location, or were changes made in support of conveying them over?

File: Bugzilla/Elastic/Role/Object.pm [hunk 2]
```diff
@@ -9,7 +9,7 @@ package Bugzilla::Elastic::Role::Object;
 use 5.10.1;
 use Role::Tiny;
 
-requires qw(ES_TYPE ES_PROPERTIES es_document);
+requires qw(ES_TYPE ES_INDEX ES_SETTINGS ES_PROPERTIES es_document);
```
Comment (floatingatoll) [inline Bugzilla/Elastic/Role/Object.pm]: To confirm, I assume this isn't meant to care about ES_PARENT_TYPE or _CLASS.

File: Bugzilla/Elastic/Indexer.pm [hunk 3]
```diff
@@ -213,7 +155,8 @@ sub _select_updated_ids {
 sub bulk_load_ids {
     my ($self, $class, $ids) = @_;
 
-    $self->put_mapping($class);
+    $self->_create_index($class);
```
Comment (floatingatoll) [inline Bugzilla/Elastic/Indexer.pm]: This was unexpected - why did you add a _create_index here? Did it not work previously?

File: Bugzilla/Elastic/Indexer.pm [hunk 4]
```diff
@@ -238,7 +181,6 @@ sub _bulk_load_ids {
     }
 
     my $total = 0;
-    use Time::HiRes;
```
Comment (floatingatoll) [inline Bugzilla/Elastic/Indexer.pm]: Can we remove all of these at once? Why is this particular ~fish~ instance deemed worthy of zapping?

File: Bugzilla/Elastic/Role/Object.pm [hunk 5]
```diff
@@ -45,4 +45,6 @@ around 'es_document' => sub {
     return $doc;
 };
 
+
```
Comment (floatingatoll) [inline Bugzilla/Elastic/Role/Object.pm]: Accidental or intentional?

File: Bugzilla/User.pm [hunk 6]
```diff
@@ -126,6 +126,10 @@ use constant EXTRA_REQUIRED_FIELDS => qw(is_enabled);
 
 with 'Bugzilla::Elastic::Role::Object';
 
+sub ES_INDEX {
+    Bugzilla->params->{elasticsearch_index} . "_user";
```
Comment (floatingatoll) [inline Bugzilla/User.pm]: `. "_" . ES_TYPE` would be a bit less repetitive of the word 'User', which is already available in the class name and then human-specified both here at +130 and below at +133 in ES_TYPE.

File: scripts/bulk_index.pl [hunk 7]
```diff
@@ -43,12 +43,10 @@ BEGIN
 }
 
 my $indexer = Bugzilla::Elastic::Indexer->new(
-    $debug_sql ? ( debug_sql => 1 ) : (),
+    $debug_sql    ? ( debug_sql => 1 )                      : (),
     $progress_bar ? ( progress_bar => 'Term::ProgressBar' ) : (),
 );
 
-$indexer->create_index;
```
Comment (floatingatoll) [inline scripts/bulk_index.pl]: It seems like this removal pairs with an above code change that auto-creates indexes now?

File: Bugzilla/Elastic/Role/Object.pm (hunk 2 above)
Comment (dylanwh) [inline Bugzilla/Elastic/Role/Object.pm]: Yep! The ChildObject role is what requires ES_PARENT_CLASS. A role is basically an interface that can provide default implementations.

File: Bugzilla/Elastic/Role/Object.pm (hunk 5 above)
Comment (dylanwh) [inline Bugzilla/Elastic/Role/Object.pm]: accidental

File: Bugzilla/Elastic/Indexer.pm (hunk 3 above)
Comment (dylanwh) [inline Bugzilla/Elastic/Indexer.pm]: Because bulk_load_ids() is public, it could be called when the index doesn't exist.

File: Bugzilla/Elastic/Indexer.pm (hunk 4 above)
Comment (dylanwh) [inline Bugzilla/Elastic/Indexer.pm]: It shouldn't have been there to begin with -- but at the top level. Sorry for the unrelated change.

File: scripts/bulk_index.pl (hunk 7 above)
Comment (dylanwh) [inline scripts/bulk_index.pl]: yes, that's correct.

File: Bugzilla/User.pm (hunk 6 above)
Comment (dylanwh) [inline Bugzilla/User.pm]: fixed in latest commit, good idea.

File: Bugzilla/Bug.pm (hunk 1 above)
Comment (dylanwh) [inline Bugzilla/Bug.pm]: Pretty much copied over. looking at this, I think I should remove two of the analyzers that are only used in the users index... but that opens the can of worms about when the index settings don't match the code -- which should be addressed in a different bug.

Comment (floatingatoll) [inline Bugzilla/Bug.pm]: Do that in a separate commit for sure :)

File: Bugzilla/Elastic/Indexer.pm (hunk 4 above)
Comment (floatingatoll) [inline Bugzilla/Elastic/Indexer.pm]: No worries, just log it in the commit message or make it a separate commit. I don't mind cleanup, I just mind *unlogged* cleanup :)

File: Bugzilla/Elastic/Indexer.pm (hunk 3 above)
Comment (floatingatoll) [inline Bugzilla/Elastic/Indexer.pm]: It would be nice if the commit message mentioned this, since no other bug exists documenting the bug.

**Comment by dylanwh** [comment]
That was merged without the proper commit message, so that commit is now gone. Re-issuing the PR.
